### ISSUE TITLE
Build from is causing error in unprivileged container

### ISSUE THREAD

**Issue by olaf-meyer** [containers/buildah#1138]
**Description**

I'm trying to build a container inside a non privileged container and I don't want to store any images on the host file system.  For this I created a centos container with buildah version 1.2 installed and 
a centos container which builds the latest buildah version 1.5-dev. When I run these container with the capability CAP_SYS_ADMIN and execute the command "buildah from" in the container I get an error message "permission denied" for buildah version 1.5-dev and "image not known" for buildah version 1.2. If I start the container as privileged than everything works fine. Do I need to give the container additional capability besides CAP_SYS_ADMIN? I attached the Dockerfiles that I'm using.

**Steps to reproduce the issue:**
1. Build container with "docker build -t build_container:centos_buildah_latest . -f Dockerfile_centos_buildah_latest" or "docker build -t build_container:centos7_buildah_1.2 . -f Dockerfile_centos_buildah_1.2"
2. Start container with "docker run --cap-add=SYS_ADMIN -it build_container:centos_buildah_latest /bin/bash" or "docker run --cap-add=SYS_ADMIN -it build_container:centos7_buildah_1.2 /bin/bash"
3. Execute inside the container the command "buildah --debug from --tls-verify=false docker.io/library/centos:7"

**Describe the results you received:**
For buildah version 1.2 "ERRO[0006] image "docker.io/library/centos:7" not found in /etc/containers/registries.conf registries: image not known" (see attachment Log_Docker_buildah_1.2_SYS_ADMIN for complete error message)
For buildah version 1.5-dev "ERRO[0006] image "ERRO[0006] Error committing the finished image: error adding layer with blob "sha256:aeb7866da422acc7e93dcf7323f38d7646f6269af33bcdb6647f2094fc4b3bf7": ApplyLayer exit status 1 stdout:  stderr: permission denied" (see attachment Log_Docker_buildah_latest_SYS_ADMIN for complete error message)

**Describe the results you expected:**
Buildah is able to download centos image from docker.io

**Output of `rpm -q buildah` or `apt list buildah`:**
Version 1.2:
```
buildah-1.2-2.gitbe87762.el7.x86_64
```
Version 1.5-dev
No possible, because buildah version was build inside the container.

**Output of `buildah version`:**
Version 1.2
```
Version:         1.2
Go Version:      go1.9.4
Image Spec:      1.0.0
Runtime Spec:    1.0.0
CNI Spec:        0.3.1
libcni Version:  v0.6.0
Git Commit:      be87762
Built:           Tue Aug 21 14:38:24 2018
OS/Arch:         linux/amd64
```

Version 1.5 
```
Version:         1.5-dev
Go Version:      go1.9.4
Image Spec:      1.0.0
Runtime Spec:    1.0.0
CNI Spec:        0.4.0
libcni Version:  v0.7.0-alpha1
Git Commit:      60485e6
Built:           Mon Oct 29 13:31:01 2018
OS/Arch:         linux/amd64
```

**Output of `podman version` if reporting a `podman build` issue:**

n/a

**Output of `cat /etc/*release`:**

```
CentOS Linux release 7.5.1804 (Core) 
```

**Output of `uname -a`:**

```
Linux 94501b4f8df2 4.13.0-45-generic #50~16.04.1-Ubuntu SMP Wed May 30 11:18:27 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

**Output of `cat /etc/containers/storage.conf`:**

```
# storage.conf is the configuration file for all tools
# that share the containers/storage libraries
# See man 5 containers-storage.conf for more information
# The "container storage" table contains all of the server options.
[storage]

# Default Storage Driver
driver = "vfs"

# Temporary storage location
runroot = "/var/run/containers/storage"

# Primary Read/Write location of container storage
graphroot = "/var/lib/containers/storage"

[storage.options]
# Storage options to be passed to underlying storage drivers

# AdditionalImageStores is used to pass paths to additional Read/Only image stores
# Must be comma separated list.
additionalimagestores = [
]

# Size is used to set a maximum size of the container image.  Only supported by
# certain container storage drivers.
size = ""

# OverrideKernelCheck tells the driver to ignore kernel checks based on kernel version
override_kernel_check = "true"

# Remap-UIDs/GIDs is the mapping from UIDs/GIDs as they should appear inside of
# a container, to UIDs/GIDs as they should appear outside of the container, and
# the length of the range of UIDs/GIDs.  Additional mapped sets can be listed
# and will be heeded by libraries, but there are limits to the number of
# mappings which the kernel will allow when you later attempt to run a
# container.
#
# remap-uids = 0:1668442479:65536
# remap-gids = 0:1668442479:65536

# Remap-User/Group is a name which can be used to look up one or more UID/GID
# ranges in the /etc/subuid or /etc/subgid file.  Mappings are set up starting
# with an in-container ID of 0 and the a host-level ID taken from the lowest
# range that matches the specified name, and using the length of that range.
# Additional ranges are then assigned, using the ranges which specify the
# lowest host-level IDs first, to the lowest not-yet-mapped container-level ID,
# until all of the entries have been used for maps.
#
# remap-user = "storage"
# remap-group = "storage"

[storage.options.thinpool]
# Storage Options for thinpool

# autoextend_percent determines the amount by which pool needs to be
# grown. This is specified in terms of % of pool size. So a value of 20 means
# that when threshold is hit, pool will be grown by 20% of existing
# pool size.
# autoextend_percent = "20"

# autoextend_threshold determines the pool extension threshold in terms
# of percentage of pool size. For example, if threshold is 60, that means when
# pool is 60% full, threshold has been hit.
# autoextend_threshold = "80"

# basesize specifies the size to use when creating the base device, which
# limits the size of images and containers.
# basesize = "10G"

# blocksize specifies a custom blocksize to use for the thin pool.
# blocksize="64k"

# directlvm_device specifies a custom block storage device to use for the
# thin pool. Required if you setup devicemapper
# directlvm_device = ""

# directlvm_device_force wipes device even if device already has a filesystem
# directlvm_device_force = "True"

# fs specifies the filesystem type to use for the base device.
# fs="xfs"

# log_level sets the log level of devicemapper.
# 0: LogLevelSuppress 0 (Default)
# 2: LogLevelFatal
# 3: LogLevelErr
# 4: LogLevelWarn
# 5: LogLevelNotice
# 6: LogLevelInfo
# 7: LogLevelDebug
# log_level = "7"

# min_free_space specifies the min free space percent in a thin pool require for
# new device creation to succeed. Valid values are from 0% - 99%.
# Value 0% disables
# min_free_space = "10%"

# mkfsarg specifies extra mkfs arguments to be used when creating the base
# device.
# mkfsarg = ""

# mountopt specifies extra mount options used when mounting the thin devices.
# mountopt = ""

# use_deferred_removal Marking device for deferred removal
# use_deferred_removal = "True"

# use_deferred_deletion Marking device for deferred deletion
# use_deferred_deletion = "True"

# xfs_nospace_max_retries specifies the maximum number of retries XFS should
# attempt to complete IO when ENOSPC (no space) error is returned by
# underlying storage device.
# xfs_nospace_max_retries = "0"

```
[Log_Docker_buildah_1.2_privileged.log](https://github.com/containers/buildah/files/2529486/Log_Docker_buildah_1.2_privileged.log)
[Log_Docker_buildah_1.2_SYS_ADMIN.log](https://github.com/containers/buildah/files/2529487/Log_Docker_buildah_1.2_SYS_ADMIN.log)
[Log_Docker_buildah_latest_privileged.log](https://github.com/containers/buildah/files/2529488/Log_Docker_buildah_latest_privileged.log)
[Log_Docker_buildah_latest_SYS_ADMIN.log](https://github.com/containers/buildah/files/2529489/Log_Docker_buildah_latest_SYS_ADMIN.log)
[Dockerfile.txt](https://github.com/containers/buildah/files/2529492/Dockerfile.txt)
[Dockerfile_centos_buildah_1.2.txt](https://github.com/containers/buildah/files/2529493/Dockerfile_centos_buildah_1.2.txt)
[Dockerfile_centos_buildah_latest.txt](https://github.com/containers/buildah/files/2529494/Dockerfile_centos_buildah_latest.txt)

**Comment by TomSweeneyRedHat** [comment]
@olaf-meyer Thanks for the issue report!  For v1.2 I'd have to guess that the container doesn't have the skopeo-containers package installed in it.  However even if you managed to get that in there, I think you'd end up running into other problems.  V1.4+ will be the right route for this as far as the privs go.  

I'll wait for @giuseppe to chime in, he'll probably have the best answer for the v1.5 error.

**Comment by giuseppe** [comment]
could you try with disabling seccomp `--security-opt seccomp=unconfined`?  I think unshare is blocked by seccomp.

**Comment by olaf-meyer** [comment]
@giuseppe I change the docker run command to:
```
docker run --cap-add=SYS_ADMIN --security-opt seccomp=unconfined -it build_container:centos_buildah_latest /bin/bash
```
Unfortunately I still get the same error message as before for both buildah versions.

**Comment by rhatdan** [comment]
Is SELinux blocking it.  

ausearch -m avc -ts recent | grep container

You could try with SELinux separation disabled.

**Comment by giuseppe** [comment]
> @giuseppe I change the docker run command to:
> 
> ```
> docker run --cap-add=SYS_ADMIN --security-opt seccomp=unconfined -it build_container:centos_buildah_latest /bin/bash
> ```
> 
> Unfortunately I still get the same error message as before for both buildah versions.

thanks for the additional information.  To get it working here with the command you've provided, I needed to specify `--isolation chroot`.  You can get the same effect with the env variable: `BUILDAH_ISOLATION=chroot`

**Comment by olaf-meyer** [comment]
@giuseppe Thanks a lot. With the new parameter it doesn't work on my local Laptop (which it running a Mint Linux and Docker 17.03). More importantly it works on a RHEL 7.5  server where I want to run the final solution. 

Anyhow, thanks a lot a again, you all helped me a great deal.

**Comment by rhatdan** [comment]
@TomSweeneyRedHat Has anyone written a blog on running buildah in a container.  Seems  a lot of people are falling over this.

**Comment by rhatdan** [comment]
Might even be good to put into docs/tutorial.

**Comment by TomSweeneyRedHat** [comment]
No blog yet, but a good subject matter.  Will get it on the list.

**Comment by olaf-meyer** [comment]
Hello,

during the weekend I had some time to play around with running buildah in a container and I can confirm that I'm able to download the base image from i.e. docker.io. If I run the command "buildah --debug bud --isolation chroot .>buildah.log" I get an error showing that the rootfs can't be mounted in the tmp directory. Has somebody seen the error ```error running subprocess: error bind mounting "/tmp/buildah281977833/hosts" from host to "/etc/hosts" in mount namespace ("/tmp/buildah281977833/mnt/rootfs/etc/hosts"): permission denied```

Is there a flag that the created directory in "/tmp" is not deleted after buildah has finished?

Thank you very much for your help in advanced.

Olaf
[buildah_bud.log](https://github.com/containers/buildah/files/2549451/buildah_bud.log)

**Comment by 42wim** [comment]
@rhatdan, @TomSweeneyRedHat  I've written a blog about that btw ;) https://blog.42.be/2018/10/buildah-inside-centos-container.html

**Comment by TomSweeneyRedHat** [comment]
Nice blog @42wim, thanks for sharing!

**Comment by rhatdan** [comment]
@olaf-meyer Can we close this issue now?

**Comment by olaf-meyer** [comment]
@rhatdan I think so. I tried to reproduce my steps as described above in a Centos virtual machine and the changes described from the blog from @42wim. If I tried to build the latest buildah version, but I get a compile error:
```
go build -ldflags '-X main.gitCommit=7c2467c -X main.buildInfo=1543502962 -X main.cniVersion=v0.7.0-alpha1'  -o buildah -tags "seccomp selinux    ostree" ./cmd/buildah
# github.com/containers/buildah/vendor/github.com/ostreedev/ostree-go/pkg/otbuiltin
vendor/github.com/ostreedev/ostree-go/pkg/otbuiltin/init.go:59: cannot use r (type C.gboolean) as type C.int in argument to isOk
vendor/github.com/ostreedev/ostree-go/pkg/otbuiltin/init.go:78: cannot use r (type C.gboolean) as type C.int in argument to isOk
make: *** [binary] Error 2
```
From my understanding this error is independent from my original issue.
